### PR TITLE
[workflow] Fix event loop can't find in thread

### LINKED ISSUE #22200: [Bug] [Workflows] Running Workflows from Python thread yields RuntimeError: There is no current event loop in thread
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Others

### What happened + What you expected to happen

Running Workflows from a Python thread is broken, see [Discuss link](https://discuss.ray.io/t/running-workflows-from-python-thread-runtimeerror-there-is-no-current-event-loop-in-thread/4861). i.e. starting a new thread and using that thread to create a fresh Ray Workflow.

Looking at the source code, I can see you plan to change how threads are created soon:

[# TODO: Get rid of this and use asyncio.run instead once we don't support py36](https://github.com/ray-project/ray/blob/ee1711fe418e44818ff791b266cd9dea42a65957/python/ray/workflow/workflow_storage.py#L58-L61)

After digging around the [workflows source](https://github.dev/ray-project/ray/blob/ee1711fe418e44818ff791b266cd9dea42a65957/python/ray/workflow) I've made progress on transitioning Workflows to `asyncio.run()` as suggested in the comment, but I could use some help.

1. Would anyone from Ray be willing to help me submit a PR to implement `asyncio.run()` threading? 

2. How I can pass a thread name to the Workflow? Where threads are created in Ray Workflows so I may set the thread name there?

A useful stack trace:
1. `connectors.py` calls `process_message()` which launches a Ray Workflow
2. Then ray calls `ray.get(self.run_async(workflow_id, metadata))` — spawning new threads
3. Ray tries to find the treads it created: `loop = asyncio.get_event_loop()`
4. Can’t find them because they [were not created by the Main Thread](https://stackoverflow.com/questions/60461433/python3-asyncio-there-is-no-current-event-loop-in-thread-when-spawn-a-new-thread) `RuntimeError: There is no current event loop in thread 'Thread-5'.`

```python
2022-01-31 10:42:23,320 [Thread-5       ] DEBUG   : pyclowder.connectors - ['kastanvday@gmail.com']
2022-01-31 10:42:23,321 [Thread-5       ] INFO    : pyclowder.connectors - [61a5853c87b65e03e7925706] : StatusMessage.start: Started processing.
2022-01-31 10:42:23,323 [Thread-5       ] DEBUG   : pyclowder.extractors - default check message : {'notifies': ['kastanvday@gmail.com'], 'source': {'id': {'resourceType': "'file", 'id': '61a5853c87b65e03e7925706'}, 'mimeType': 'text/plain', 'extra': {'filename': '10chars.txt'}}, 'jobid': '61f82d8e193c9580af4aa3bf', 'msgid': '61f82d8e193c9580af4aa3c0', 'flags': '', 'intermediateId': '61a5853c87b65e03e7925706', 'host': 'http://host.docker.internal:9000', 'datasetId': '61a41c945e5855829e5923d8', 'id': '61a5853c87b65e03e7925706', 'fileSize': '11', 'filename': '10chars.txt', 'target': '{}', 'secretKey': '70c420d3-918a-45d9-a4e3-987f726c6694', 'activity': 'submitted', 'routing_key': 'extractors.ncsa.rayWorkflow', 'parameters': {}, 'action': 'manual-submission', 'retry_count': 1}
2022-01-31 10:42:23,370 [Thread-5       ] INFO    : pyclowder.connectors - [61a5853c87b65e03e7925706] : StatusMessage.processing: Downloading file.
2022-01-31 10:42:23,461 [Thread-5       ] INFO    : pyclowder.connectors - [61a5853c87b65e03e7925706] : StatusMessage.processing: Loading contents of file...
10:42:23.462680
2022-01-31 10:42:23,479	INFO execution.py:49 -- Workflow job created. [id="rayWorkflow-10:42:23.462680", storage_url="file:///tmp/data"]. Type: FUNCTION.
2022-01-31 10:42:23,481 [Thread-5       ] ERROR   : pyclowder.connectors - [61a5853c87b65e03e7925706] There is no current event loop in thread 'Thread-5'.
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/pyclowder/connectors.py", line 426, in _process_message
    self.process_message(self, source_host, secret_key, resource, body)
  File "rayWorkflow.py", line 68, in process_message
    output = self.workflow.run(f"rayWorkflow-{current_time}")
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/workflow/common.py", line 363, in run
    return ray.get(self.run_async(workflow_id, metadata))
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/workflow/common.py", line 405, in run_async
    return run(self, workflow_id, metadata)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/workflow/execution.py", line 56, in run
    ws.save_workflow_user_metadata(metadata)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/workflow/workflow_storage.py", line 416, in save_workflow_user_metadata
    self._put(self._key_workflow_user_metadata(), metadata, True))
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/workflow/workflow_storage.py", line 55, in asyncio_run
    loop = asyncio.get_event_loop()
  File "/home/ray/anaconda3/lib/python3.7/asyncio/events.py", line 644, in get_event_loop
    % threading.current_thread().name)
RuntimeError: There is no current event loop in thread 'Thread-5'.
2022-01-31 10:42:23,483 [Thread-5       ] INFO    : pyclowder.connectors - [61a5853c87b65e03e7925706] : StatusMessage.retry: (#2) There is no current event loop in thread 'Thread-5'.
```

Thank you!

### Versions / Dependencies

* Ray 1.9
* Python 3.8
* Ubuntu

Another library creates a thread (via asyncio) and calls my Ray Workflow. That library is [PyClowder](https://github.com/clowder-framework/pyclowder) v2.4.1.

### Reproduction script

```python
from ray import workflow
from typing import List

@workflow.step
def fake_data(num: int):
    return [i for i in range(num)]

# Start new thread here ⚠️
import threading

class thread(threading.Thread):
    def __init__(self, thread_name, thread_ID):
        threading.Thread.__init__(self)
        self.thread_name = thread_name
        self.thread_ID = thread_ID
 
        # helper function to execute the threads
    def run(self):
        print(str(self.thread_name) +" "+ str(self.thread_ID))
        # Initialize workflow storage.
        workflow.init()

        # Setup the workflow.
        data = fake_data.step(10)
        print(data.run())

# main 
thread1 = thread("thread_outside_ray", 42)
 
thread1.start()

# [Output err] RuntimeError: There is no current event loop in thread
```

### Anything else

I'd love to help fix this! I'm a grad student at UIUC, and we are part of the IBM Accelerator Institute which is contributing to Ray Workflows (particularly [Ray DAG](https://github.com/ray-project/contrib-workflow-dag)). If you can help me troubleshoot, I'd love to submit a PR!

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Event loop will only be set in main thread by default and this will make workflow unable to work if it's called in thread other than main thread which can happen when it's called from a library (for example ray serve).
This PR fixed it.

## Related issue number
Closes https://github.com/ray-project/ray/issues/22200

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/workflow/common.py [hunk 1]
```diff
@@ -23,6 +25,18 @@
 STORAGE_ACTOR_NAME = "StorageManagementActor"
 
 
+def asyncio_run(coro):
+    # The event loop will only be set in main thread by default
```
Comment (stephanie-wang) [inline python/ray/workflow/common.py:29]: Maybe clean up this comment?

**Review by stephanie-wang — APPROVED** [review]

**Comment by iycheng** [comment]
@ericl  any comment about this?

File: python/ray/workflow/common.py [hunk 2]
```diff
@@ -23,6 +25,15 @@
 STORAGE_ACTOR_NAME = "StorageManagementActor"
 
 
+def asyncio_run(coro):
+    try:
+        loop = asyncio.get_event_loop()
+    except RuntimeError:
```
Comment (ericl) [inline python/ray/workflow/common.py:31]: Is there a cleaner way that detect this?

**Review by ericl — APPROVED** [review]

File: python/ray/workflow/common.py (hunk 2 above)
Comment (ericl) [reply python/ray/workflow/common.py:31]: Hmm guess not: https://docs.python.org/3/library/asyncio-eventloop.html

**Pull request merged**